algorithms/kernel/svm/oneapi/svm_train_thunder_oneapi_impl.i

```diff
@@ -0,0 +1,277 @@
+services::Status SVMTrainOneAPI<algorithmFPType, ParameterType, thunder>::updateGrad(const services::Buffer<algorithmFPType> & kernelWS,
+                                                                                     const services::Buffer<algorithmFPType> & deltaalpha,
+                                                                                     services::Buffer<algorithmFPType> & grad, const size_t nVectors,
+                                                                                     const size_t nWS)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(updateGrad);
+    return BlasGpu<algorithmFPType>::xgemm(math::Layout::RowMajor, math::Transpose::Trans, math::Transpose::NoTrans, nVectors, 1, nWS,
+                                           algorithmFPType(1), kernelWS, nVectors, 0, deltaalpha, 1, 0, algorithmFPType(1), grad, 1, 0);
+}
+
+template <typename algorithmFPType, typename ParameterType>
+services::Status SVMTrainOneAPI<algorithmFPType, ParameterType, thunder>::smoKernel(
+    const services::Buffer<algorithmFPType> & y, const services::Buffer<algorithmFPType> & kernelWsRows, const services::Buffer<uint32_t> & wsIndices,
+    const uint32_t ldK, const services::Buffer<algorithmFPType> & f, const algorithmFPType C, const algorithmFPType eps, const algorithmFPType tau,
+    const uint32_t maxInnerIteration, services::Buffer<algorithmFPType> & alpha, services::Buffer<algorithmFPType> & deltaalpha,
+    services::Buffer<algorithmFPType> & resinfo, const size_t nWS)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(smoKernel);
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+    auto & factory = context.getClKernelFactory();
+
+    services::String build_options = getKeyFPType<algorithmFPType>();
+
+    services::String cachekey("__daal_algorithms_svm_smo_block_");
+    build_options.add(" -D WS_SIZE=");
+    char bufferString[DAAL_MAX_STRING_SIZE] = { 0 };
+    services::daal_int_to_string(bufferString, DAAL_MAX_STRING_SIZE, int(nWS));
+    build_options.add(bufferString);
+    build_options.add(" -D SIMD_WIDTH=64 ");
+    cachekey.add(build_options);
+
+    services::Status status;
+    factory.build(ExecutionTargetIds::device, cachekey.c_str(), clKernelBlockSMO, build_options.c_str(), &status);
+
+    DAAL_CHECK_STATUS_VAR(status);
+
+    auto kernel = factory.getKernel("smoKernel");
+
+    KernelArguments args(12);
+    args.set(0, y, AccessModeIds::read);
+    args.set(1, kernelWsRows, AccessModeIds::read);
+    args.set(2, wsIndices, AccessModeIds::read);
+    args.set(3, ldK);
+    args.set(4, f, AccessModeIds::read);
+    args.set(5, C);
+    args.set(6, eps);
+    args.set(7, tau);
+    args.set(8, maxInnerIteration);
+    args.set(9, alpha, AccessModeIds::readwrite);
+    args.set(10, deltaalpha, AccessModeIds::readwrite);
+    args.set(11, resinfo, AccessModeIds::readwrite);
+
+    KernelRange localRange(nWS);
+    KernelRange globalRange(nWS);
```

Comment:
Does it mean that it always be only one work-group (GPU Gen9/Gen12 subslice) performing the work, and the rest of GPU is idle?

Reply:
Yes, But this part, as it turns out, works fast enough. If we use fully HW -  the time required for synchronization will be huge. I think for the multiclass it will be possible to think about how to win from this. Because we will train independently `K*(K-1)/2` models.